compiler/noirc_evaluator/src/ssa/opt/unrolling.rs

```diff
@@ -529,166 +971,402 @@ mod tests {
+        assert_eq!(loops.yet_to_unroll.len(), 1);
+
+        let (lower, upper) = loops.yet_to_unroll[0]
+            .get_const_bounds(function, &loops.cfg)
+            .expect("should find bounds")
+            .expect("bounds are numeric const");
+
+        assert_eq!(lower, FieldElement::from(0u32));
+        assert_eq!(upper, FieldElement::from(4u32));
+    }
 
-        let v0 = builder.add_parameter(Type::field());
-        let v1 = builder.add_block_parameter(b1, Type::field());
+    #[test]
+    fn test_find_pre_header_reference_values() {
+        let ssa = brillig_unroll_test_case();
+        let function = ssa.main();
+        let mut loops = Loops::find_all(function);
+        let loop0 = loops.yet_to_unroll.pop().unwrap();
+
+        let refs = loop0.find_pre_header_reference_values(function, &loops.cfg).unwrap();
+        assert_eq!(refs.len(), 1);
+        assert!(refs.contains(&ValueId::new(2)));
+
+        let (loads, stores) = loop0.count_loads_and_stores(function, &refs);
+        assert_eq!(loads, 1);
+        assert_eq!(stores, 1);
+
+        let all = loop0.count_all_instructions(function);
+        assert_eq!(all, 7);
+    }
 
-        builder.terminate_with_jmp(b1, vec![v0]);
+    #[test]
+    fn test_boilerplate_stats() {
+        let ssa = brillig_unroll_test_case();
+        let stats = loop0_stats(&ssa);
+        assert_eq!(stats.iterations, 4);
+        assert_eq!(stats.all_instructions, 2 + 5); // Instructions in b1 and b3
+        assert_eq!(stats.increments, 1);
+        assert_eq!(stats.loads, 1);
+        assert_eq!(stats.stores, 1);
+        assert_eq!(stats.useful_instructions(), 1); // Adding to sum
+        assert_eq!(stats.baseline_instructions(), 8);
+        assert!(stats.is_small());
+    }
 
-        builder.switch_to_block(b1);
-        let five = builder.field_constant(5u128);
-        let v2 = builder.insert_binary(v1, BinaryOp::Lt, five);
-        builder.terminate_with_jmpif(v2, b2, b3);
+    #[test]
+    fn test_boilerplate_stats_6470() {
+        let ssa = brillig_unroll_test_case_6470(3);
+        let stats = loop0_stats(&ssa);
+        assert_eq!(stats.iterations, 3);
+        assert_eq!(stats.all_instructions, 2 + 8); // Instructions in b1 and b3
+        assert_eq!(stats.increments, 2);
+        assert_eq!(stats.loads, 1);
+        assert_eq!(stats.stores, 1);
+        assert_eq!(stats.useful_instructions(), 3); // array get, add, array set
+        assert_eq!(stats.baseline_instructions(), 11);
+        assert!(stats.is_small());
+    }
 
-        builder.switch_to_block(b2);
-        let one = builder.field_constant(1u128);
-        let v3 = builder.insert_binary(v1, BinaryOp::Add, one);
-        builder.terminate_with_jmp(b1, vec![v3]);
+    /// Test that we can unroll a small loop.
+    #[test]
+    fn test_brillig_unroll_small_loop() {
+        let ssa = brillig_unroll_test_case();
+
+        // Example taken from an equivalent ACIR program (ie. remove the `unconstrained`) and run
```

Comment:
Confused by this comment. Do you mean to point the reader to a specific program? Also they should remove the `unconstrained` even though main is a brillig runtime?

Reply:
The program is in the comments of the [brillig_unroll_test_case](https://github.com/noir-lang/noir/blob/176d0fca2a4e8c58a8ef2eb257e1834b6097dbf6/compiler/noirc_evaluator/src/ssa/opt/unrolling.rs#L1293) below. I just wanted to explain where this expected SSA came from: by treating the same code as ACIR and capturing the output, and to trigger the ACIR workflow I removed `unconstrained` from `main`. I'll rephrase it.

The nargo command is probably not needed for anyone used to working with these though, I'll remove it.